include/alp/base/internalops.hpp

```diff
@@ -31,3144 +31,198 @@
-			 * definition as a GraphBLAS operator with the #has_foldl, #has_foldr, and
-			 * the other fields, set as required.
-			 *
-			 * @tparam IN1 The left-hand input data type.
-			 * @tparam IN2 The right-hand input data type.
-			 * @tparam OUT The output data type.
+			 * @see grb::operators::internal::left_assign
 			 */
 			template< typename IN1, typename IN2, typename OUT, enum Backend implementation = config::default_backend >
-			class left_assign {
-
-			public:
-				/** Alias to the left-hand input data type. */
-				typedef IN1 left_type;
-
-				/** Alias to the right-hand input data type. */
-				typedef IN2 right_type;
-
-				/** Alias to the output data type. */
-				typedef OUT result_type;
-
-				/** Whether this operator has an inplace foldl. */
-				static constexpr bool has_foldl = true;
-
-				/** Whether this operator has an inplace foldr. */
-				static constexpr bool has_foldr = true;
-
-				/**
-				 * Whether this operator is \em mathematically associative; that is,
-				 * associative when assuming equivalent data types for \a IN1, \a IN2,
-				 * and \a OUT, as well as assuming exact arithmetic, no overflows, etc.
-				 */
-				static constexpr bool is_associative = true;
-
-				/**
-				 * Whether this operator is \em mathematically commutative; that is,
-				 * commutative when assuming equivalent data types for \a IN1, \a IN2,
-				 * and \a OUT, as well as assuming exact arithmetic, no overflows, etc.
-				 */
-				static constexpr bool is_commutative = false;
-
-				/**
-				 * Out-of-place application of the addition c = a.
-				 *
-				 * @param[in]  a The left-hand side input. Must be pre-allocated and initialised.
-				 * @param[in]  b The right-hand side input. Must be pre-allocated and initialised.
-				 * @param[out] c The output. Must be pre-allocated.
-				 */
-				static void apply( const left_type * __restrict__ const a, const right_type * __restrict__ const b, result_type * __restrict__ const c ) {
-					(void)b;
-					*c = static_cast< result_type >( *a );
-				}
-
-				/**
-				 * In-place left-to-right folding.
-				 *
-				 * @param[in]     a Pointer to the left-hand side input data.
-				 * @param[in,out] c Pointer to the right-hand side input data. This also
-				 *                  dubs as the output memory area.
-				 */
-				static void foldr( const left_type * __restrict__ const a, result_type * __restrict__ const c ) {
-					*c = static_cast< result_type >( *a );
-				}
-
-				/**
-				 * In-place right-to-left folding.
-				 *
-				 * @param[in,out] c Pointer to the left-hand side input data. This also
-				 *                  dubs as the output memory area.
-				 * @param[in]     b Pointer to the right-hand side input data.
-				 */
-				static void foldl( result_type * __restrict__ const c, const right_type * __restrict__ const b ) {
-					(void)b;
-					(void)c;
-				}
-			};
+			using left_assign = grb::operators::internal::left_assign< IN1, IN2, OUT, grb::Backend::reference >;
```

Comment:
Not sure if it is a good idea to stick to `grb::Backend::reference` at this point. what if we are in a non-reference scenario and the operator is also specialized in grb? This would not go for the right specialization right?
If this comment makes sense then should we have a thin mapping layer `alp::Backend::x` -> `grb::Backend::y` ?

Similar remark for a number of other occurrences in this file and other non-reference headers.